docs/configuration.md
# 🎛 Configuration

Lightz Client can be configured via a `TOML` config file or CLI arguments. By
default, the config file is called `lightz.toml` and located in the data
directory of Lightz Client (`~/.lightzd` by default on Linux).

## Example

```toml
# Path to the log file
logfile = ""

# possible values: fatal, error, warn, info, debug, silly
loglevel = "info"

# possible values: "mainnet", "testnet" or "regtest"
network = "mainnet"

# you will have to set this to "cln" or "lnd" if you have configuration values for both
node = ""

# Whether to use Lightz Pro fee rates
pro = false

# The daemon can also operate without a lightning node
# standalone = true

# Custom referral ID to use when creating swaps
referralId = "my-referral"

[LIGHTNING]
# Default fee limit in ppm for lightning payments. Can be overridden on a per-swap basis.
routingFeeLimitPpm = 2500

[BOLTZ]
# By default the daemon automatically connects to the official Lightz Backend for the network your node is on
# This value is used to overwrite that
# url = "https://api.lightz.exchange"

[DATABASE]
# Path to the SQLite database file
# path = "~/test.db"

[LND]
# Host of the gRPC interface of LND
# host = "127.0.0.1"

# Port of the gRPC interface of LND
# port = 10009

# Path to the data directory of LND
# datadir = "~/.lnd"

# Path to a macaroon file of LND.
# The daemon needs to have permission to read various endpoints, generate addresses and pay invoices
# Not required if datadir is specified
# macaroon = "~/.lnd/data/chain/bitcoin/mainnet/admin.macaroon"

# Path to the TLS certificate of LND
# Not required if datadir is specified
# certificate = "~/.lnd/tls.cert"

[CLN]
# Host of the gRPC interface of CLN
# host = "127.0.0.1"

# Port of the gRPC interface of CLN
# port = 9736

# Path to the data directory of CLN
# datadir = "~/.lightning"

# Paths to TLS certificates and keys of CLN. Not required if datadir is specified
# rootcert = "~/.lightning/bitcoin/ca.pem"
# privatekey = "~/.lightning/bitcoin/client-key.pem"
# certchain =  "~/.lightning/bitcoin/client.pem"

[RPC]
# Host of the gRPC interface
host = "127.0.0.1"

# Port of the gRPC interface
port = 9002

# Whether the REST proxy for the gRPC interface should be disabled
restDisabled = false

# Host of the REST proxy
restHost = "127.0.0.1"

# Port of the REST proxy
restPort = 9003

# Path to the TLS cert for the gRPC and REST interface
tlsCert = ""

# Path to the TLS private key for the gRPC and REST interface
tlsKey = ""

# Whether the macaroon authentication for the gRPC and REST interface should be disabled
noMacaroons = false

# Path to the admin macaroon for the gRPC and REST interface
adminMacaroonPath = ""

# Path to the read-only macaroon for the gRPC and REST interface
readOnlyMacaroonPath = ""

# Password for authentication (alternative to macaroons and takes precedence if set)
password = ""
```


internal/config/config.go
package config

import (
	"fmt"
	"os"
	"path"
	"runtime"
	"strings"

	"github.com/lightzapp/lightz-client/internal/logger"
	"github.com/lightzapp/lightz-client/internal/onchain"
	"gopkg.in/natefinch/lumberjack.v2"

	"github.com/BurntSushi/toml"
	"github.com/jessevdk/go-flags"

	"github.com/lightzapp/lightz-client/internal/build"
	"github.com/lightzapp/lightz-client/internal/cln"
	"github.com/lightzapp/lightz-client/internal/database"
	"github.com/lightzapp/lightz-client/internal/lnd"
	"github.com/lightzapp/lightz-client/internal/utils"
)

type helpOptions struct {
	ShowHelp    bool `short:"h" long:"help" description:"Display this help message"`
	ShowVersion bool `short:"v" long:"version" description:"Display version and exit"`
}

type LightzOptions struct {
	URL string `long:"lightz.url" description:"Lightz API URL"`
}

type RpcOptions struct {
	Host string `long:"rpc.host" description:"gRPC host to which Lightz should listen"`
	Port int    `long:"rpc.port" short:"p" description:"gRPC port to which Lightz should listen"`

	RestHost     string `long:"rpc.rest.host" description:"REST host to which Lightz should listen"`
	RestPort     int    `long:"rpc.rest.port" description:"REST port to which Lightz should listen"`
	RestDisabled bool   `long:"rpc.rest.disable" description:"Disables the REST API proxy"`

	TlsCertPath string `long:"rpc.tlscert" description:"Path to the TLS certificate of lightzd-client"`
	TlsKeyPath  string `long:"rpc.tlskey" description:"Path to the TLS private key of lightzd-client"`
	NoTls       bool   `long:"rpc.no-tls" description:"Disables TLS"`

	NoMacaroons          bool   `long:"rpc.no-macaroons" description:"Disables Macaroon authentication"`
	AdminMacaroonPath    string `long:"rpc.adminmacaroonpath" description:"Path to the admin Macaroon"`
	ReadonlyMacaroonPath string `long:"rpc.readonlymacaroonpath" description:"Path to the readonly macaroon"`
	Password             string `long:"rpc.password" description:"Password for authentication (alternative to macaroons and takes precedence if set)" json:"-"`
}

type LightningOptions struct {
	RoutingFeeLimitPpm uint64 `long:"lightning.routing-fee-limit-ppm" description:"Default fee limit in ppm for lightning payments. Can be overridden on a per-swap basis."`
}

type Config struct {
	DataDir string `short:"d" long:"datadir" description:"Data directory of lightzd-client"`

	ConfigFile string `short:"c" long:"configfile" description:"Path to configuration file"`

	LogFile    string `short:"l" long:"logfile" description:"Path to the log file"`
	LogLevel   string `long:"loglevel" description:"Log level (fatal, error, warn, info, debug, silly)"`
	LogMaxSize int    `long:"logmaxsize" description:"Maximum size of the log file in megabytes before it gets rotated"`
	LogMaxAge  int    `long:"logmaxage" description:"Maximum age of old log files in days before they get deleted"`

	Log logger.Options `json:"-"`

	Network string `long:"network" description:"Network to use (mainnet, testnet, regtest)"`

	Lightz *LightzOptions `group:"Lightz Options"`
	LND    *lnd.LND       `group:"LND Options"`
	Cln    *cln.Cln       `group:"Cln Options"`

	Node string `long:"node" description:"Lightning node to use (cln or lnd)"`

	Standalone bool `long:"standalone" description:"Run lightzd-client without a lightning node"`

	Lightning *LightningOptions  `group:"Lightning options"`
	RPC       *RpcOptions        `group:"RPC options"`
	Database  *database.Database `group:"Database options"`

	MempoolApi       string `long:"mempool" description:"mempool.space API to use for fee estimations; set to empty string to disable"`
	MempoolLiquidApi string `long:"mempool-liquid" description:"mempool.space liquid API to use for fee estimations; set to empty string to disable"`

	ElectrumUrl       string `long:"electrum" description:"electrum rpc to use for fee estimations; set to empty string to disable"`
	ElectrumSSL       bool   `long:"electrum-ssl" description:"whether the electrum server uses ssl"`
	ElectrumLiquidUrl string `long:"electrum-liquid" description:"electrum rpc to use for fee estimations; set to empty string to disable"`
	ElectrumLiquidSSL bool   `long:"electrum-liquid-ssl" description:"whether the electrum server uses ssl"`

	Proxy string `long:"proxy" description:"Proxy URL to use for all Lightz API requests"`

	Pro        bool   `long:"pro" description:"Use the Lightz Pro API"`
	ReferralId string `long:"referral-id" description:"Custom referral ID to use when creating swaps"`

	MaxZeroConfAmount        *uint64 `long:"max-zeroconf-amount" description:"Maximum amount of sats to accept 0-conf"`
	AutoConsolidateThreshold *uint64 `long:"auto-consolidate-threshold" description:"Number of UTXOs that trigger auto consolidation. Set to 0 to disable"`

	Help *helpOptions `group:"Help Options"`
}

func (c *Config) Electrum() onchain.ElectrumConfig {
	config := onchain.ElectrumConfig{}
	if c.ElectrumUrl != "" {
		config.Btc = &onchain.ElectrumOptions{Url: c.ElectrumUrl, SSL: c.ElectrumSSL}
	}
	if c.ElectrumLiquidUrl != "" {
		config.Liquid = &onchain.ElectrumOptions{Url: c.ElectrumLiquidUrl, SSL: c.ElectrumLiquidSSL}
	}
	return config
}

func LoadConfig(dataDir string) (*Config, error) {
	cfg := Config{
		DataDir: dataDir,

		ConfigFile: "",

		LogLevel:   "info",
		LogMaxSize: 5,
		LogMaxAge:  30,

		Network: "mainnet",

		Lightz: &LightzOptions{},

		LND: &lnd.LND{
			Host:        "127.0.0.1",
			Port:        10009,
			Macaroon:    "",
			Certificate: "",
		},

		Cln: &cln.Cln{
			Host: "127.0.0.1",
			Port: 10009,

			RootCert:   "",
			PrivateKey: "",
			CertChain:  "",
			ServerName: "cln",
		},

		Lightning: &LightningOptions{
			RoutingFeeLimitPpm: 2500,
		},

		RPC: &RpcOptions{
			Host: "127.0.0.1",
			Port: 9002,

			RestHost:     "127.0.0.1",
			RestPort:     9003,
			RestDisabled: false,

			TlsCertPath: "",
			TlsKeyPath:  "",

			NoMacaroons:          false,
			AdminMacaroonPath:    "",
			ReadonlyMacaroonPath: "",
		},

		Database: &database.Database{
			Path: "",
		},

		ReferralId: "lightzd-client",
	}

	parser := flags.NewParser(&cfg, flags.IgnoreUnknown)
	_, err := parser.Parse()
	if err != nil {
		printCouldNotParse(err)
	}

	if cfg.Help.ShowVersion {
		fmt.Println(build.GetVersion())
		fmt.Println("Built with: " + runtime.Version())
		os.Exit(0)
	}

	if cfg.Help.ShowHelp {
		parser.WriteHelp(os.Stdout)
		os.Exit(0)
	}

	cfg.ConfigFile = utils.ExpandDefaultPath(cfg.DataDir, cfg.ConfigFile, "lightz.toml")

	if cfg.ConfigFile != "" && utils.FileExists(cfg.ConfigFile) {
		_, err := toml.DecodeFile(cfg.ConfigFile, &cfg)

		if err != nil {
			return nil, fmt.Errorf("could not read config file: %v", err)
		}
	}

	// parse a second time to ensure cli flags go over config values
	_, err = parser.Parse()
	if err != nil {
		printCouldNotParse(err)
	}

	fmt.Println("Using data dir: " + cfg.DataDir)

	if strings.EqualFold(cfg.Node, "CLN") && cfg.Cln.DataDir == "" {
		cfg.Cln.DataDir = "~/.lightning"
	} else if strings.EqualFold(cfg.Node, "LND") && cfg.LND.DataDir == "" {
		cfg.LND.DataDir = "~/.lnd"
	}

	cfg.LND.Macaroon = utils.ExpandHomeDir(cfg.LND.Macaroon)
	cfg.LND.Certificate = utils.ExpandHomeDir(cfg.LND.Certificate)

	if cfg.LND.DataDir != "" {
		cfg.LND.DataDir = utils.ExpandHomeDir(cfg.LND.DataDir)
		if cfg.Network == "" {
			return nil, fmt.Errorf("network must be set when lnd datadir is configured")
		}

		defaultMacaroon := fmt.Sprintf("./data/chain/bitcoin/%s/admin.macaroon", cfg.Network)
		cfg.LND.Macaroon = utils.ExpandDefaultPath(cfg.LND.DataDir, cfg.LND.Macaroon, defaultMacaroon)
		cfg.LND.Certificate = utils.ExpandDefaultPath(cfg.LND.DataDir, cfg.LND.Certificate, "tls.cert")
	}

	cfg.Cln.RootCert = utils.ExpandHomeDir(cfg.Cln.RootCert)
	cfg.Cln.PrivateKey = utils.ExpandHomeDir(cfg.Cln.PrivateKey)
	cfg.Cln.CertChain = utils.ExpandHomeDir(cfg.Cln.CertChain)

	if cfg.Cln.DataDir != "" {
		cfg.Cln.DataDir = utils.ExpandHomeDir(cfg.Cln.DataDir)
		if cfg.Network == "" {
			return nil, fmt.Errorf("network must be set when cln datadir is configured")
		}
		if cfg.Network == "mainnet" {
			cfg.Cln.DataDir += "/bitcoin"
		} else {
			cfg.Cln.DataDir += "/" + cfg.Network
		}
		cfg.Cln.RootCert = utils.ExpandDefaultPath(cfg.Cln.DataDir, cfg.Cln.RootCert, "ca.pem")
		cfg.Cln.PrivateKey = utils.ExpandDefaultPath(cfg.Cln.DataDir, cfg.Cln.PrivateKey, "client-key.pem")
		cfg.Cln.CertChain = utils.ExpandDefaultPath(cfg.Cln.DataDir, cfg.Cln.CertChain, "client.pem")
	}

	cfg.LogFile = utils.ExpandDefaultPath(cfg.DataDir, cfg.LogFile, "lightz.log")
	cfg.Log = logger.Options{
		Level: cfg.LogLevel,
		Logger: &lumberjack.Logger{
			Filename: cfg.LogFile,
			MaxAge:   cfg.LogMaxAge,
			MaxSize:  cfg.LogMaxSize,
		},
	}
	cfg.Database.Path = utils.ExpandDefaultPath(cfg.DataDir, cfg.Database.Path, "lightz.db")

	cfg.RPC.TlsKeyPath = utils.ExpandDefaultPath(cfg.DataDir, cfg.RPC.TlsKeyPath, "tls.key")
	cfg.RPC.TlsCertPath = utils.ExpandDefaultPath(cfg.DataDir, cfg.RPC.TlsCertPath, "tls.cert")

	macaroonDir := path.Join(cfg.DataDir, "macaroons")

	cfg.RPC.AdminMacaroonPath = utils.ExpandDefaultPath(macaroonDir, cfg.RPC.AdminMacaroonPath, "admin.macaroon")
	cfg.RPC.ReadonlyMacaroonPath = utils.ExpandDefaultPath(macaroonDir, cfg.RPC.ReadonlyMacaroonPath, "readonly.macaroon")

	if cfg.Pro {
		cfg.ReferralId = "pro"
	}

	createDirIfNotExists(cfg.DataDir)
	createDirIfNotExists(macaroonDir)

	return &cfg, nil
}

func createDirIfNotExists(dir string) {
	if !utils.FileExists(dir) {
		err := os.Mkdir(dir, 0700)

		if err != nil {
			fmt.Println("Could not create directory: " + err.Error())
			os.Exit(1)
		}
	}
}

func printCouldNotParse(err error) {
	fmt.Println("Could not parse arguments: " + err.Error())
	os.Exit(1)
}
